docs/source/specs/typespec/main.tsp

```diff
@@ -0,0 +1,251 @@
+import "@typespec/http";
+import "@typespec/rest";
+import "@typespec/openapi3";
+import "@typespec/versioning";
+
+using TypeSpec.Http;
+using TypeSpec.Rest;
+using TypeSpec.Versioning;
+
+@service({title:"Role Based Access Control for Kessel Project"})
+
+@server("https://localhost:{port}/{basePath}", "Development Server", { basePath: string="api/rbac/v2", port: string="8000" })
+@server("https://console.stage.redhat.com/{basePath}", "Stage Server", { basePath: string="api/rbac/v2" })
+@server("https://console.redhat.com/{basePath}", "Production Server", { basePath: string="api/rbac/v2" })
+@versioned(Versions)
+
+namespace KesselRBAC;
+
+enum Versions{
+    v2,
+}
+
+model Link {
+    @format("uri")
+    first: string = "/api/v2/(resources)/?offset=0&limit=10";
+    @format("uri")
+    previous: string = "/api/v2/(resources)/?offset=20&limit=10";
+    @format("uri")
+    next: string = "/api/v2/(resources)/?offset=40&limit=10";
+    @format("uri")
+    last: string = "/api/v2/(resources)/?offset=90&limit=10";
+}
+
+model Meta {
+    count: int64 = 30
+}
+
+model ListResponse<T> {
+    data: T[];
+    links: Link;
+    meta: Meta;
+}
+
+model Timestamps {
+    @doc("2019-01-21T17:32:28Z")
+    created: offsetDateTime;
+    @doc("2019-01-21T17:32:28Z")
+    modified: offsetDateTime;
+}
+
+model ListQueryParams {
+    @doc("Parameter for selecting the amount of data returned.")
+    @maxValue(1000)
+    @minValue(1)
+    @query limit?: integer = 10,
+    @query offset?: integer = 0,
+    @doc("Parameter for ordering by value. For inverse ordering, supply '-' before the param value, such as: ?order_by=-name")
+    @query order_by?: OrderByDefaultValues;
+}
+
+@format("uuid")
+scalar UUID extends string;
+
+enum OrderByDefaultValues {
+    name,
+    modified
+}
+
+
+model OkResponseForModel<T>{
+    @statusCode
+    statusCode: 200,
+    ...T
+}
+
+
+model CreatedResponseForModel<T>{
+    @statusCode
+    statusCode: 201,
+    ...T
+}
+
+/**
+ * Resource Deleted
+ */
+model NoContentResponseResourceDeleted{
+    @statusCode
+    statusCode: 201,
+}
+
+namespace Errors {
```

Comment:
Would it be possible to make our error objects comply with https://www.rfc-editor.org/rfc/rfc9457.html ?

If possible it would be nice to define a generic type which models this, and allows extension to define the schema of additional fields.

I sketched out something like this:

```
@error
model Problem<T extends ProblemDetails> {
  ...T,
}

enum ProblemType {
  "https://project-kessel.org/problems/oops",
}

model ProblemDetails {
  type: ProblemType,
  status: int16,
  title?: string,
  detail?: string,
}

model Oopsy extends ProblemDetails {
  type: ProblemType.`https://project-kessel.org/problems/oops`,
  oopsy: string,
  @statusCode
  status: 500,
  title: "There ways an oopsy",
}
```

There is an open upstream issue to include something native in Typespec for this which may be slated for August: https://github.com/microsoft/typespec/issues/2553

As long as what we did is compatible, shouldn't matter in the meantime.

Reply:
Done 👍